packages/web-components/src/components/va-modal/va-modal.tsx

```diff
@@ -0,0 +1,304 @@
+  styleUrl: 'va-modal.css',
+  shadow: true,
+})
+export class VaModal {
+  alertActions: HTMLDivElement;
+  closeButton: HTMLButtonElement;
+  contents: HTMLDivElement;
+  isVisibleDirty: boolean; // used to determine if visible has changed (workaround)
+  focusTrap: FocusTrap;
+  undoAriaHidden: Undo; // used to undo aria hidden on close
+
+  @Element() el: HTMLElement;
+
+  /**
+   * Fires when modal is closed
+   */
+  @Event({
+    composed: true,
+    bubbles: true,
+  })
+  closeEvent: EventEmitter;
+
+  /**
+   * The event used to track usage of the component. Fires when a
+   * a page is selected if enable-analytics is true.
+   */
+  @Event({
+    eventName: 'component-library-analytics',
+    composed: true,
+    bubbles: true,
+  })
+  componentLibraryAnalytics: EventEmitter;
+
+  /**
+   * Click outside modal will trigger closeEvent
+   */
+  @Prop() clickToClose?: boolean = false;
+
+  /**
+   * If true, analytics event won't be fired
+   */
+  @Prop() disableAnalytics?: boolean = false;
+
+  /**
+   * Title/header text for the modal
+   */
+  @Prop() modalTitle?: string;
+
+  /**
+   * Hide the close button that's normally in the top right
+   */
+  @Prop() hideCloseButton?: boolean = false;
+
+  /**
+   * Selector to explicitly specify which element should receive
+   * focus when the modal is open, if the initially focused element
+   * is not the first focusable element in the document
+   */
+  @Prop() initialFocusSelector?: string;
+
+  /**
+   * Primary button text and action
+   */
+  @Prop() primaryButton?: { text: string; action: () => void };
+
+  /**
+   * Secondary button text and action
+   */
+  @Prop() secondaryButton?: { text: string; action: () => void };
+
+  /*
+   * Style of modal alert - info, error, success, warning
+   */
+  @Prop() status?: 'continue' | 'error' | 'info' | 'success' | 'warning';
+
+  /**
+   * If the modal is visible or not
+   */
+  @Prop() visible: boolean = false;
+
+  @Listen('click')
+  handleClick(e) {
+    if (!this.clickToClose) return;
+
+    // event.target is always the shadow host
+    // event.composedPath()[0] returns the node clicked when shadow root is open
+    if (this.visible && e.composedPath()[0] === this.el) {
+      this.handleClose(e);
+    }
+  }
+
+  @Listen('keydown', { target: 'window' })
+  handleKeyDown(e: KeyboardEvent) {
+    if (!this.visible) return;
+
+    const keyCode = e.key;
+    if (keyCode === 'Escape') {
+      this.handleClose(e);
+    }
+  }
+
+  @Watch('visible')
+  watchVisibleHandler() {
+    // This is a workaround for determining when to call setupModal or teardownModal.
+    // Elements are not yet available in the DOM due to `if (!visible) return null;`
+    // See componentDidUpdate.
+    this.isVisibleDirty = true;
+  }
+
+  componentDidLoad() {
+    if (this.visible) this.setupModal();
+  }
+
+  componentDidUpdate() {
+    // This dirty check prevents the component from updating endlessly.
+    if (!this.isVisibleDirty) return;
```

Comment:
This is needed to provide the same functionality as React's componentDidUpdate. Stencil's `componentDidUpdate` doesn't provide previous prop values that we can use to decide if `setupModal` or `teardownModal` should be invoked. We can use `@Watch` to know when `visible` changes and perform the check in `componentDidUpdate` to prevent endless updating.